test/libsolidity/ASTPropertyTest.cpp

```diff
@@ -0,0 +1,202 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+// SPDX-License-Identifier: GPL-3.0
+
+#include <test/libsolidity/ASTPropertyTest.h>
+#include <test/Common.h>
+
+#include <libsolidity/ast/ASTJsonExporter.h>
+#include <libsolidity/interface/CompilerStack.h>
+
+#include <liblangutil/Common.h>
+#include <liblangutil/SourceReferenceFormatter.h>
+#include <libsolutil/JSON.h>
+
+#include <boost/algorithm/string.hpp>
+#include <boost/throw_exception.hpp>
+
+#include <range/v3/algorithm/find_if.hpp>
+#include <range/v3/range/conversion.hpp>
+#include <range/v3/view/split.hpp>
+
+#include <queue>
+
+using namespace solidity::util;
+using namespace solidity::langutil;
+using namespace solidity::frontend;
+using namespace solidity::frontend::test;
+using namespace solidity;
+using namespace std;
+
+ASTPropertyTest::ASTPropertyTest(string const& _filename):
+	TestCase(_filename)
+{
+	if (!boost::algorithm::ends_with(_filename, ".sol"))
+		BOOST_THROW_EXCEPTION(runtime_error("Not a Solidity file: \"" + _filename + "\"."));
+
+	m_source = m_reader.source();
+	readExpectations();
+}
+
+string ASTPropertyTest::formatExpectations(bool _obtainedResult)
+{
+	string expectations;
+	for (string const& testId: m_testOrder)
+	{
+		soltestAssert(m_tests.count(testId) > 0);
+		expectations +=
+			testId +
+			": " +
+			(_obtainedResult ? m_tests[testId].obtainedValue : m_tests[testId].expectedValue)
+			+ "\n";
+	}
+	return expectations;
+}
+
+vector<StringPair> ASTPropertyTest::readKeyValuePairs(string const& _input)
+{
+	vector<StringPair> result;
+	for (string const& inputLine: _input | ranges::views::split('\n') | ranges::to<vector<string>>)
+	{
+		string line = boost::trim_copy(inputLine);
```

Comment:
This avoids having a line with only spaces being processed further.
I was trying to do this as a step of the `ranges` pipeline in the `for` loop above, but I'll have to learn how to do that first :sweat_smile:

Reply:
I don't think `ranges` has `trim()` so this is the way to do it currently :)

Reply:
Actually, maybe you could use `transform()` with `trim_copy` as the function. This will still probably end up less concise than this.

One thing you could change though would be make `inputLine` a `string` (rather than a reference) and then use normal `trim()` in place. `to<>` has to create a copy anyway. But that's just a microoptimization :)

Reply:
> Actually, maybe you could use transform() with trim_copy as the function. This will still probably end up less concise than this.

I looked at some examples and docs and was considering using `transform`, but from what I learned, I also think it would end up being less concise.

> One thing you could change though would be make inputLine a string [...]

I will change that. It's fairly simple.